### PR TITLE
[1.20.x] Pull Upstream Changes from MinecraftForge/MinecraftForge

### PR DESCRIPTION
**Pull request by Shadows-of-Fire**
This PR pulls all changes up through [MinecraftForge 47.1.28](https://maven.minecraftforge.net/net/minecraftforge/forge/1.20.1-47.1.28/forge-1.20.1-47.1.28-changelog.txt)

Importantly - do not squash and merge, we want to keep the commit history for these.

### CODE REVIEW

File: src/main/java/net/minecraftforge/common/ForgeConfig.java [hunk 1]
```diff
@@ -127,7 +135,7 @@ public static class Client {
                     .define("useCombinedDepthStencilAttachment", false);
 
             compressLanIPv6Addresses = builder
-                    .comment("When enabled, Forge will convert discovered 'Open to LAN' IPv6 addresses to their more compact, compressed representation")
+                    .comment("[DEPRECATED] [NOW IN COMMON CONFIG] When enabled, Forge will convert discovered 'Open to LAN' IPv6 addresses to their more compact, compressed representation")
```
Comment (pupnewfster) [inline src/main/java/net/minecraftforge/common/ForgeConfig.java:138]: This doesn't seem to be in the common config currently but does seem to be deprecated/unused? @PaintNinja Can you explain should this comment just have the deprecated bit and not the now in common config part?

File: src/main/java/net/minecraftforge/common/extensions/IForgeItemStack.java [hunk 2]
```diff
@@ -292,11 +292,21 @@ default float getXpRepairRatio()
...
+    @Deprecated(forRemoval = true, since = "1.20.1") // Use onInventoryTick
     default void onArmorTick(Level level, Player player)
     {
         self().getItem().onArmorTick(self(), level, player);
     }
 
+    /**
+     * Called to tick this items in a players inventory, the indexes are the global slot index.
+     */
+    default void onInventoryTick(Level level, Player player, int slotIndex, int selectedIndex)
```
Comment (pupnewfster) [inline src/main/java/net/minecraftforge/common/extensions/IForgeItemStack.java:304]: This is not currently hooked up

File: src/main/java/net/minecraftforge/event/entity/EntityEvent.java [hunk 3]
```diff
@@ -140,48 +141,93 @@ public boolean didChunkChange()
...
+     * This event is fired whenever {@link IForgeEntity#getEyeHeightForge(Pose)} gets called.<br>
+     * CAREFUL: This is also fired in the Entity constructor. Therefore the entity(subclass) might not be fully initialized. Check Entity#isAddedToWorld() or !Entity#firstUpdate.<br>
+     * <br>
+     * This event is not {@link Cancelable}.<br>
+     * <br>
+     * This event does not have a result. {@link HasResult}
+     * <br>
+     * This event is fired on the {@link MinecraftForge#EVENT_BUS}.<br>
+     **/
+    public static class EyeHeight extends EntityEvent
```
Comment (pupnewfster) [inline src/main/java/net/minecraftforge/event/entity/EntityEvent.java:231]: Some methods being moved to this new event from the size event is a breaking change as any references to them on the size event previously no longer exist and will cause method not found errors.

File: src/main/java/net/minecraftforge/common/extensions/IForgeItemStack.java (hunk 2 above)
Comment (Shadows-of-Fire) [reply src/main/java/net/minecraftforge/common/extensions/IForgeItemStack.java:304]: I'm not a fan of this existing at all, tbh - the signature is virtually the same to the vanilla one and it would be more realistic to just fix the index.  I realize we have to have it for bincompat, but yeah.

Comment (pupnewfster) [reply src/main/java/net/minecraftforge/common/extensions/IForgeItemStack.java:304]: Worse than that the way/spot it would be hooked into... would pass a different value for slotIndex than the implementation of this method expects so we would have to modify it decently before calling it.

Comment (Shadows-of-Fire) [reply src/main/java/net/minecraftforge/common/extensions/IForgeItemStack.java:304]: Since we have to keep this method for ABI, do we want to deprecate it, call it anyway, and just proxy to the vanilla method with the correct index? I need to scrub vanilla overrides and see if any of them actually call it, but the way the call currently is means the `slotIndex` is not actually usable, since you don't know which sub-inventory it came from.

File: patches/minecraft/net/minecraft/world/entity/player/Inventory.java.patch [hunk 4]
```diff
@@ -19,8 +19,18 @@
           if (p_36049_.m_41782_()) {
              itemstack.m_41751_(p_36049_.m_41783_().m_6426_());
           }
-@@ -229,7 +_,7 @@
+@@ -222,14 +_,16 @@
+    }
+ 
+    public void m_36068_() {
++      int slotIndex = 0; // Neo - Provide the "proper" slot index as correlated with Inventory#getItem
```
Comment (Matyrobbrt) [inline patches/minecraft/net/minecraft/world/entity/player/Inventory.java.patch]: Probably worth noting that the purpose of the patch is Forge compatibility

Comment (Shadows-of-Fire) [reply patches/minecraft/net/minecraft/world/entity/player/Inventory.java.patch]: Only partially - this _is_ an actual bug fix for providing the proper slot index, the forge-compat part is calling the deprecated `IForgeItemStack#onInventoryTick` method instead of calling `ItemStack#inventoryTick` directly.

File: src/main/java/net/minecraftforge/event/entity/EntityEvent.java (hunk 3 above)
Comment (Shadows-of-Fire) [reply src/main/java/net/minecraftforge/event/entity/EntityEvent.java:231]: Yeah this is a pretty major API breakage. Undoing the break while keeping forge-parity is a PITA though, because we have to undo all the changes to Size and somehow make it work alongside EyeHeight

Comment (Matyrobbrt) [reply src/main/java/net/minecraftforge/event/entity/EntityEvent.java:231]: Seems like there definitely are mods affected by the change (see https://github.com/MinecraftForge/MinecraftForge/issues/9659)

Comment (pupnewfster) [reply src/main/java/net/minecraftforge/event/entity/EntityEvent.java:231]: At the very least we may want to add the methods back deprecate for removal (maybe even add error logging on call) so that mods who didn't test functionality (due to said event having not actually been fired) won't straight up crash, even if they will still be in the situation of pre the events being fixed and firing again of not having expected functionality due to them not testing their mod. (And then with the logging users could actually report said issues to the mods and hopefully get them to fix it and just reference the new event)

File: src/main/java/net/minecraftforge/common/extensions/IForgeItemStack.java (hunk 2 above)
Comment (pupnewfster) [reply src/main/java/net/minecraftforge/common/extensions/IForgeItemStack.java:304]: I am not sure how valid a solution that is as I [mentioned some on discord](https://discord.com/channels/313125603924639766/852298000042164244/1130154262421323847), given I definitely have used the "faulty" index for purposes of checking if something is on the hotbar or offhand but not in main inventory. Which the changes/decision with this would break (while the mod I do this on isn't on 1.20.1 yet, there may be other mods that do similar, and at the very least it would not be particularly evident that the way the index worked changed during porting)

File: patches/minecraft/net/minecraft/world/entity/player/Inventory.java.patch (hunk 4 above)
Comment (pupnewfster) [reply patches/minecraft/net/minecraft/world/entity/player/Inventory.java.patch]: Would this be a decent time though to decide on how we want to do NeoForge comments? Basically decide on the following:
1. Do we want a space between the comment and the word Neo?
2. Do we want to do like this with a `Neo - <details>` or do we want to do something like `Neo: <details>`?
3. Do we want to use `Neo` or `NeoForge`?

File: src/main/java/net/minecraftforge/common/ForgeConfig.java (hunk 1 above)
Comment (PaintNinja) [reply src/main/java/net/minecraftforge/common/ForgeConfig.java:138]: Spoke with Lex, it's now always-on with no option to turn it off. The original PR intended on merging the compression for logging and LAN discovery into a single config option.

The reason for there being an option to turn off IP compression for LAN discovery was a "just in case" scenario where I wanted to give people something to try if they came to me with any problems and it turned out to be IPv6 LAN working *except* when using compressed addresses. It's been a long time and that's never been an issue for anyone, so Lex changed it to be always-on.

So basically: debugging option deemed unnecessary after a long time of unuse -> new unified option for logging and debug lan discovery got removed from the PR before merge -> old config option remains in case someone programmatically reads that setting and expects it to exist but no longer has any effect.

Rename it to `"[DEPRECATED] [NO EFFECT]"`

Comment (Shadows-of-Fire) [reply src/main/java/net/minecraftforge/common/ForgeConfig.java:138]: Wouldn't we just remove it then?

Comment (PaintNinja) [reply src/main/java/net/minecraftforge/common/ForgeConfig.java:138]: > old config option remains in case someone programmatically reads that setting and expects it to exist but no longer has any effect

File: patches/minecraft/net/minecraft/world/entity/player/Inventory.java.patch (hunk 4 above)
Comment (Shadows-of-Fire) [reply patches/minecraft/net/minecraft/world/entity/player/Inventory.java.patch]: I think I like `Neo: <stuff>` as a format
